feat: Settings + Import Module

Takshil-Kunadia opened the pull request:
- Add settings page.
- Add Import parser.
- Add User create/update logic.
- Add background processing for import.
- Add cron job for scheduling these processes.

Takshil-Kunadia committed:
feat(print-circ): add dependency checker and settings management

Takshil-Kunadia committed:
feat(print-circ): implement CSV import functionality and initializer class

Takshil-Kunadia committed:
feat(print-circ): enhance settings class with user role management options

Takshil-Kunadia committed:
feat(print-circ): add Newspack Fields class for field mapping

Takshil-Kunadia committed:
feat(print-circ): add Import Parser & refactor user import process to follow updated flow & lint code

Takshil-Kunadia commented on includes/class-initializer.php:
```diff
@@ -0,0 +1,107 @@
+	public static function temp_process_csv() {
+		$import_module = new Newspack_Print_Import();
+		// phpcs:ignore Squiz.PHP.CommentedOutCode.Found
+		// $import_module->fetch_csv_file();
```
Currently, for testing I had hooked it on 'init'. This will be removed and the import will be done via cron or cli

Takshil-Kunadia commented on includes/class-settings.php:
```diff
@@ -0,0 +1,385 @@
+		if ( isset( $input[ self::IMPORT_TRANSFORMATIONS_OPTION ] ) ) {
+			// Check if the input is a valid callable.
+			$import_transformations = $input[ self::IMPORT_TRANSFORMATIONS_OPTION ];
+			/**
+			 * TODO: eval is a huge security risk! need to find a better way to do this.
+			 */
+			$import_transformations = eval( 'return ' . $import_transformations . ';' ); // phpcs:ignore Squiz.PHP.Eval.Discouraged
```
This is also a challenge as of yet.

Takshil-Kunadia commented on includes/class-import-parser.php:
```diff
@@ -0,0 +1,53 @@
+
+		$import_transformations = Settings::get_setting( Settings::IMPORT_TRANSFORMATIONS_OPTION );
+		if ( ! empty( $import_transformations ) ) {
+			/**
+			 * TODO: eval is a huge security risk! need to find a better way to do this.
+			 */
+			$import_transformations = eval( 'return ' . $import_transformations . ';' ); // phpcs:ignore Squiz.PHP.Eval.Discouraged
```
This is also a challenge as of yet.

leogermani replied:
yeah, I was thinking about creating a list of forbidden strings, to avoid the most obvious issues.. like WP functions that mess with the database, filesystem, base64decode/encode etc...

But it's ok to have some level of risk. It's like all of those WP plugins that allow you to add code snippets to your site

Takshil-Kunadia committed:
feat(print-circ): enable CSV file fetching logic

Takshil-Kunadia committed:
refactor(print-circ): enhance CSV import functionality with error handling and cleanup methods

Takshil-Kunadia committed:
fix(print-circ): offset handling and resetting file pointer

Takshil-Kunadia committed:
fix(import): change file mode to 'w+' for temporary CSV file creation

Takshil-Kunadia committed:
feat(import): implement background processing for user import in batches

Takshil-Kunadia committed:
fix(import): optimise import process

Takshil-Kunadia committed:
fix(import): only fetch CSV when required

Takshil-Kunadia committed:
fix(print-circ): scope wp-background-process dependencies to resolve package conflicts

Takshil-Kunadia committed:
feat(import-cron): Implement cron job scheduling for user import/export synchronization

Takshil-Kunadia committed:
feat(fields): add status field for user status

Takshil-Kunadia commented on includes/class-newspack-fields.php:
```diff
@@ -0,0 +1,203 @@
+		$field = self::get_field( $slug );
+
+		// Set the field value based on the field type.
+		if ( ! empty( $field['type'] ) && 'user_meta' === $field['type'] ) {
```
Sanitisation required here too!

Takshil-Kunadia commented on includes/class-newspack-fields.php:
```diff
@@ -0,0 +1,203 @@
+		$field = self::get_field( $slug );
+
+		// Set the field value based on the field type.
+		if ( ! empty( $field['type'] ) && 'user_meta' === $field['type'] ) {
+			// Store as a user meta field.
+			update_user_meta( $user_id, $field['db_field'], $value );
+		} elseif ( ! empty( $field['type'] ) && 'user_prop' === $field['type'] ) {
+			// Store as a user property.
+			$user = get_user_by( 'id', $user_id );
+			$user->{$field['db_field']} = $value;
+			wp_update_user( $user );
+		} else {
+			$extra_field = self::get_field( self::EXTRA_FIELD );
+			// Store as a part of the user's extra data.
+			$extra_data = get_user_meta( $user_id, $extra_field['db_field'], true );
+			if ( ! is_array( $extra_data ) ) {
+				$extra_data = [];
+			}
+			$extra_data[ $slug ] = $value;
+			update_user_meta( $user_id, $extra_field['db_field'], $extra_data );
```
Sanitisation required here too!

Takshil-Kunadia replied:
Also currently it saves empty for blank csv values, we can choose to not save empty values.

Takshil-Kunadia committed:
fix(import): skip rows with incomplete information

Takshil-Kunadia committed:
fix(settings): handle undefined option values in text and textarea fields

Takshil-Kunadia committed:
feat(logging): implement logging functionality with settings and log management

leogermani commented:
Hi @Takshil-Kunadia overall this is looking good and is on the right path!

But I think we are working on too many fronts at the same time and it makes it difficult to test and debug. I suggest we focus only on the import process now, done directly via WP Shell or a CLI command, and leave the other aspects of it for later:

* No need to support CSV files living outside the server. The path can be to a file in the server and you can simply use `file_get_contents`. No need to read remote CSV files.
* I think UI is only adding complexity at this point. Let's work with a configuration file and let's hardcode a defeault configuration file with what we need to work with the sample CSV we have.
* Let's leave CRON and Background process for later. Let's test things by calling them directly via WP shell

You could focus on

* Add a logging class: So we can have detailed logs of what's going on - either writing to a file or outputting in the terminal when calling via CLI
* Add a `test_import` method that will only parse the file and run the mapping on the top 10 rows. Just so we can test the transformation and tweak the mapping and the code snippets without writing anything to the DB

Let me know if it makes sense to you

Takshil-Kunadia committed:
feat(import): enhance user import process with batch tracking and logging

Takshil-Kunadia committed:
fix(settings): improved layout

Takshil-Kunadia committed:
feat(cli): add WP-CLI command for user import process

Takshil-Kunadia committed:
feat(cli): refactor user import process to process users synchronously

Takshil-Kunadia committed:
feat(import): enhance user import process with validation and logging of skipped users

Takshil-Kunadia committed:
feat(cli): add dry run option to user import process for safe testing

Takshil-Kunadia committed:
feat(cli): limit user processing to first batch during dry run mode

Takshil-Kunadia committed:
feat(cli): reduce batch size to 5 users during dry run mode and update output messages

Takshil-Kunadia commented:
@leogermani 
- Added a logging class that saves the latest 20 entries and displays it at the Logs sub-menu of Newspack Print Settings Admin menu.

<img width="1301" alt="Screenshot 2025-04-03 at 4 28 15 PM" src="https://github.com/user-attachments/assets/a8e53693-92e6-4fe3-99a8-6e4da6935046" />

- Also added the test_import_users function that won't save anything to DB (https://github.com/Automattic/newspack-print-circ-integration/pull/2/commits/6e91dd3d16022b5e48d52404a149621a0b40b4bf)

Takshil-Kunadia committed:
feat(import): add test import functionality for user batches

leogermani committed:
many tweaks on importer. settings and mapping

Takshil-Kunadia committed:
refactor(import): streamline user import process by removing unused code

Takshil-Kunadia committed:
feat(access): implement access manager for user membership handling

Takshil-Kunadia committed:
fix(settings): enhance display name generation logic for missing info users

Takshil-Kunadia committed:
fix(access): remove duplication in granting access

Takshil-Kunadia committed:
feat(access): update existing membership status

Takshil-Kunadia committed:
fix(access): update text domain

Takshil-Kunadia committed:
feat(export): add parser for parsing export user data

Takshil-Kunadia committed:
feat(export): add CSV exporter for user data

Takshil-Kunadia committed:
fix(settings): correct display names and improve address & status parsing logic

Takshil-Kunadia committed:
feat(export): add WP-CLI command for exporting users to CSV

Takshil-Kunadia committed:
feat(export): detail logging for user export process

Takshil-Kunadia committed:
feat(import): enhance membership status updates with descriptive messages

Takshil-Kunadia committed:
fix(settings): disable subscription settings

Takshil-Kunadia committed:
feat(export): add dry-run option for user export

Takshil-Kunadia committed:
fix(settings): correct import & export transformations

Takshil-Kunadia committed:
fix(export): add log message for empty user export scenario

Takshil-Kunadia committed:
feat(import): enhance user import functionality with extra data handling

Takshil-Kunadia committed:
feat(import/export): save CSV fields for csv fields order consistency

leogermani committed:
Merge pull request #4 from Automattic/feature/cli-store-all-data

feat: Store All data

leogermani committed:
fix: remove unused methods and improve logs

leogermani committed:
chore: mostly linting and one adjust in logs

leogermani committed:
Merge pull request #3 from Automattic/feature/cli-commands

Feature/cli commands
